state/protocol/badger/mutator.go

```diff
@@ -1091,25 +1061,30 @@ func (m *FollowerState) versionBeaconOnBlockFinalized(
 //     operations to insert service events for blocks that include them.
 //
 // No errors are expected during normal operation.
-func (m *FollowerState) handleEpochServiceEvents(candidate *flow.Block) (dbUpdates []func(*transaction.Tx) error, err error) {
+func (m *FollowerState) handleEpochServiceEvents(candidate *flow.Block, updater protocol.StateUpdater) (dbUpdates []func(*transaction.Tx) error, err error) {
 	epochFallbackTriggered, err := m.isEpochEmergencyFallbackTriggered()
 	if err != nil {
 		return nil, fmt.Errorf("could not retrieve epoch fallback status: %w", err)
 	}
-	epochStatus, err := m.epochStatus(candidate.Header, epochFallbackTriggered)
-	if err != nil {
-		return nil, fmt.Errorf("could not determine epoch status for candidate block: %w", err)
-	}
-	activeSetup, err := m.epoch.setups.ByID(epochStatus.CurrentEpoch.SetupID)
-	if err != nil {
-		return nil, fmt.Errorf("could not retrieve current epoch setup event: %w", err)
-	}
+	parentProtocolState := updater.ParentState()
+	epochStatus := parentProtocolState.EpochStatus()
+	activeSetup := parentProtocolState.CurrentEpochSetup
 
-	// always persist the candidate's epoch status
-	// note: We are scheduling the operation to store the Epoch status using the _pointer_ variable `epochStatus`.
-	// The struct `epochStatus` points to will still be modified below.
-	blockID := candidate.ID()
-	dbUpdates = append(dbUpdates, m.epoch.statuses.StoreTx(blockID, epochStatus))
+	// perform protocol state transition to next epoch if next epoch is committed and we are at first block of epoch
+	if !epochFallbackTriggered {
+		phase, err := epochStatus.Phase()
+		if err != nil {
+			return nil, fmt.Errorf("could not determine epoch phase: %w", err)
+		}
+		if phase == flow.EpochPhaseCommitted {
+			if candidate.Header.View > parentProtocolState.CurrentEpochSetup.FinalView {
+				err = updater.TransitionToNextEpoch()
+				if err != nil {
+					return nil, fmt.Errorf("could not transition protocol state to next epoch: %w", err)
+				}
+			}
+		}
+	}
```

Comment:
My gut feeling:
* It would be beneficial to avoid a change of the Protocol state at the epoch switchover. 
* While this is technically permitted in [our formal model](https://www.notion.so/dapperlabs/Dynamic-Protocol-State-Meeting-Notes-Sep-6-2023-b30937cdbe1c44759fc4cf0a1b414478?pvs=4#327f59e0ec324a9cb3aab0149a8a7e3b), I don't think it is necessary

Suggestion:
* for this PR, I would leave the code as is and add a TODO
* This point here is already part of our issue https://github.com/onflow/flow-go/issues/4649